otherlibs/win32unix/socketpair.c

```diff
@@ -0,0 +1,202 @@
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*                         Antonin Decimo, Tarides                        */
+/*                                                                        */
+/*   Copyright 2021 Tarides                                               */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#include <caml/mlvalues.h>
+#include <caml/memory.h>
+#include <caml/alloc.h>
+#include <caml/misc.h>
+#include <caml/signals.h>
+#include "unixsupport.h"
+#include <errno.h>
+
+#ifdef HAS_SOCKETS
+
+#include "socketaddr.h"
+#include <ws2tcpip.h>
+
+extern int socket_domain_table[]; /* from socket.c */
+extern int socket_type_table[]; /* from socket.c */
+
+#ifdef HAS_SOCKETPAIR
+
+#error "Windows has defined sockepair! win32unix should be updated."
+CAMLprim value unix_socketpair(value cloexec, value domain, value type,
+                               value protocol)
+{ caml_invalid_argument("socketpair not implemented"); }
+
+#else
+
+static int socketpair(int domain, int type, int protocol,
+                      SOCKET socket_vector[2])
+{
+  wchar_t dirname[MAX_PATH + 1], path[MAX_PATH + 1];
+  union sock_addr_union addr;
+  socklen_param_type socklen;
+
+  /* POSIX states that in case of error, the contents of socket_vector
+     shall be unmodified. */
+  SOCKET listener = INVALID_SOCKET,
+    server = INVALID_SOCKET,
+    client = INVALID_SOCKET;
+
+  fd_set writefds, exceptfds;
+  u_long non_block, peerid = 0UL;
+
+  DWORD drc;
+  int rc;
+
+  if (GetTempPath(MAX_PATH + 1, dirname) == 0) {
+    win32_maperr(GetLastError());
+    goto fail;
+  }
+
+  if (GetTempFileName(dirname, L"osp", 0U, path) == 0) {
+    win32_maperr(GetLastError());
+    goto fail;
+  }
+
+  addr.s_unix.sun_family = PF_UNIX;
+  socklen = sizeof(addr.s_unix);
+
+  /* sun_path needs to be set in UTF-8 */
+  rc = WideCharToMultiByte(CP_UTF8, 0, path, -1, addr.s_unix.sun_path,
+                           UNIX_PATH_MAX, NULL, NULL);
+  if (rc == 0) {
+    win32_maperr(GetLastError());
+    goto fail_path;
+  }
+
+  listener = socket(domain, type, protocol);
+  if (listener == INVALID_SOCKET)
+    goto fail_wsa;
+
+  /* The documentation requires removing the file before binding the socket. */
+  if (DeleteFile(path) == 0) {
+    win32_maperr(GetLastError());
+    if (errno != ENOENT)
+      goto fail_sockets;
```

Comment:
```suggestion
    drc = GetLastError();
    if (drc != ERROR_FILE_NOT_FOUND) {
      win32_maperr(drc);
      goto fail_sockets;
    }
```
(the present codes leaves `errno` in a strange state if the file was otherwise deleted)